File: TextKit/src/commonMain/kotlin/SearchTerms.kt
package antoni.kalorie.textkit

fun searchTerms(input: String): List<String> =
    foldDiacritics(input.lowercase())
        .split(Regex(" +"))
        .filter { it.isNotEmpty() }
        .flatMap { word -> (1..word.length).map { length -> word.take(length) } }
        .distinct()

data class SearchQuery(val lowercased: String, val folded: String, val lastWord: String)

fun searchQuery(query: String): SearchQuery {
    val lowercased = query.lowercase()
    val folded = foldDiacritics(lowercased)
    val lastWord = folded.split(" ").lastOrNull { it.isNotEmpty() } ?: folded
    return SearchQuery(lowercased = lowercased, folded = folded, lastWord = lastWord)
}

File: scripts/lib/search-terms.js
const { foldDiacritics } = require('./diacritics');

function searchTerms(input) {
  const folded = foldDiacritics(input.toLowerCase());
  const terms = new Set();
  for (const word of folded.split(/ +/).filter((w) => w.length > 0)) {
    for (let length = 1; length <= word.length; length++) {
      terms.add(word.slice(0, length));
    }
  }
  return Array.from(terms);
}

module.exports = { searchTerms };
